### DISCUSSION TITLE
agent调用多次后会报错

### DISCUSSION THREAD

**Discussion by ElevenPP** [zai-org/ChatGLM3#307] [Bad Case]
### System Info / 系統信息

Package                   Version 
------------------------- ------------ 
accelerate                0.24.1       
aiofiles                  23.2.1       
aiohttp                   3.8.6        
aiosignal                 1.3.1        
altair                    5.1.2        
annotated-types           0.6.0        
anyio                     3.7.1        
arxiv                     2.0.0
asttokens                 2.4.1
async-timeout             4.0.3
attrs                     23.1.0
blinker                   1.7.0
cachetools                5.3.2
certifi                   2023.7.22
charset-normalizer        3.3.2
click                     8.1.7
colorama                  0.4.6
comm                      0.2.0
contourpy                 1.2.0
cpm-kernels               1.0.11
cycler                    0.12.1
dataclasses-json          0.6.1
debugpy                   1.8.0 
decorator                 5.1.1
distro                    1.8.0
exceptiongroup            1.1.3
executing                 2.0.1
fastapi                   0.95.1
feedparser                6.0.10
ffmpy                     0.3.1
filelock                  3.13.1
flit_core                 3.9.0
fonttools                 4.44.0
frozenlist                1.4.0
fsspec                    2023.10.0
gitdb                     4.0.11
GitPython                 3.1.40
gradio                    3.39.0
gradio_client             0.7.0
greenlet                  3.0.1
h11                       0.14.0
httpcore                  1.0.1
httpx                     0.25.1
huggingface-hub           0.18.0
idna                      3.4
importlib-metadata        6.8.0
ipykernel                 6.26.0
ipython                   8.17.2
jedi                      0.19.1
Jinja2                    3.1.2 
jsonpatch                 1.33
jsonpointer               2.4
jsonschema                4.19.2
jsonschema-specifications 2023.7.1
jupyter_client            8.6.0
jupyter_core              5.5.0
kiwisolver                1.4.5
langchain                 0.0.330
langsmith                 0.0.57
latex2mathml              3.76.0
linkify-it-py             2.0.2
loguru                    0.5.3
Markdown                  3.5.1
markdown-it-py            2.2.0
MarkupSafe                2.1.3
marshmallow               3.20.1
matplotlib                3.8.1
matplotlib-inline         0.1.6
mdit-py-plugins           0.3.3
mdtex2html                1.2.0
mdurl                     0.1.2
mpmath                    1.3.0
multidict                 6.0.4
mypy-extensions           1.0.0
nest-asyncio              1.5.8
networkx                  3.2.1
numpy                     1.26.1
openai                    0.28.0
orjson                    3.9.10
packaging                 23.2
pandas                    2.1.2
parso                     0.8.3
Pillow                    10.1.0
pip                       23.3
platformdirs              3.11.0
prompt-toolkit            3.0.39
protobuf                  4.25.0
psutil                    5.9.6
pure-eval                 0.2.2
pyarrow                   14.0.0
pydantic                  1.10.13
pydantic_core             2.10.1
pydeck                    0.8.1b0
pydub                     0.25.1
Pygments                  2.16.1
pyparsing                 3.1.1
python-dateutil           2.8.2
python-multipart          0.0.6
pytz                      2023.3.post1
pywin32                   306
PyYAML                    6.0.1
pyzmq                     25.1.1
referencing               0.30.2 
regex                     2023.10.3
requests                  2.31.0
rich                      13.6.0
rpds-py                   0.12.0
safetensors               0.4.0
semantic-version          2.10.0
sentencepiece             0.1.99
setuptools                68.0.0
sgmllib3k                 1.0.0
six                       1.16.0
smmap                     5.0.1
sniffio                   1.3.0
SQLAlchemy                2.0.23
sse-starlette             1.6.5
stack-data                0.6.3
starlette                 0.26.1
streamlit                 1.28.1
sympy                     1.12
tenacity                  8.2.3
tokenizers                0.13.3
toml                      0.10.2
toolz                     0.12.0
torch                     2.1.0+cu121
torchaudio                2.1.0+cu121
torchvision               0.16.0+cu121
tornado                   6.3.3 
tqdm                      4.66.1
traitlets                 5.13.0
transformers              4.30.2
typing_extensions         4.4.0
typing-inspect            0.9.0
tzdata                    2023.3
tzlocal                   5.2
uc-micro-py               1.0.2
urllib3                   2.0.7
uvicorn                   0.24.0
validators                0.22.0
watchdog                  3.0.0
wcwidth                   0.2.9
websockets                11.0.3
wheel                     0.41.2
win32-setctime            1.1.0
yarl                      1.9.2
zipp                      3.17.0

### Information / 问题信息

- [X] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

在tool页面多次自动使用查询天气工具get_weather，几次过后会报错Failed to parse tool call

### Expected behavior / 期待表现

能正常调用

**Comment by zRzRzRzRzRzRzR** [comment]
是连续调用，还是开启新的对话报错

**Comment by cl886699** [comment]
> 是连续调用，还是开启新的对话报错

同样的问题，有时候问第二次就出现问题了[
![image](https://github.com/THUDM/ChatGLM3/assets/31495593/d5626a2b-26fa-4c37-8643-4d587784cc70)
](url)
还有一个问题，问天气不会去调用工具[
![image](https://github.com/THUDM/ChatGLM3/assets/31495593/85e15a8d-7be1-4139-89cf-5ba7bc269033)
](url)
我使用调用工具的感受是，一般第一次比较准确，经过多轮对话后的回答就不靠谱了，即使是只使用同一个工具

**Comment by tai960519073** [comment]
我使用讯飞3.0的Tool功能也发现这个问题，第一次还行，后面就有很高概率胡扯

**Comment by zRzRzRzRzRzRzR** [comment]
> 我使用讯飞3.0的Tool功能也发现这个问题，第一次还行，后面就有很高概率胡扯

我们的demo没有办法适配友商的模型

**Comment by zRzRzRzRzRzRzR** [comment]
> > 是连续调用，还是开启新的对话报错
> 
> 同样的问题，有时候问第二次就出现问题了[ ](url)![image](https://github.com/THUDM/ChatGLM3/assets/31495593/d5626a2b-26fa-4c37-8643-4d587784cc70) 还有一个问题，问天气不会去调用工具[ ](url)![image](https://github.com/THUDM/ChatGLM3/assets/31495593/85e15a8d-7be1-4139-89cf-5ba7bc269033) 我使用调用工具的感受是，一般第一次比较准确，经过多轮对话后的回答就不靠谱了，即使是只使用同一个工具

这应该是一个bad case，能够复现
能够解决的办法目前是调优提示词
可以到Discussion看看有没有对应的相似讨论，我们会在discussion记录讨论的bad case并在之后的模型优化上调优

**Reply by si1mon2** [reply]
遇到同样的问题，无论是get_weather还是自己定义新的function，第一次都可以调用，但是多次调用就不起作用了，要么提示Failed to parse tool call，要么提示已经执行了，但是并没有实质的调用。不知道是不是history里面append了之前的conversation，对后续的多次调用产生了影响。

**Comment by zRzRzRzRzRzRzR** [comment]
这是一个bad case，我们能够复现，可能需要在未来对模型进行优化
